<font size=5>
Finding the sum of the off diagonal terms of a matrix. 
</font>

In [1]:
import numpy as np

In [2]:
A = np.asmatrix( [[1,2,3], [4,5,6], [7,8,9]] )
B = np.asmatrix( [[1,2,3,4], [5,6,7,8], [9,10,11,12], [13,14,15,16]] )

Before starting, we'd like to know the solutions to both matrices, so we know if we are correct.

In [3]:
np.sum( [2,3,4,6,7,8] ), np.sum( [2,3,4,5,7,8,9,10,12,13,14,15] )

(30, 102)

In [4]:
from functools import reduce
reduce(lambda x,y: x+y, [2,3,4,6,7,8]), reduce(lambda x,y: x+y, [2,3,4,5,7,8,9,10,12,13,14,15])

(30, 102)

Answers are:
    * 30 for matrix A
    * 102 for matrix B

```

```
__My first attempt at it was this:__

In [5]:
def sum_of_off_diags(mat):
    off_diags = []
    for i,row in enumerate(mat):
        row = row.tolist()[0]
        for j,element in enumerate(row):
            if i==j:
                pass
            else:
                off_diags.append(element)
    return np.sum(off_diags)

In [6]:
sum_of_off_diags(A)

30

In [7]:
sum_of_off_diags(B)

102

Although this works, it's not the quickest way. However, it tends to be my instinct to first attack a problem using for loops and to use as few of the imported modules as possible. In this one, I did not even use the `numpy` module.

```

```
__My second attempt.__

I admit, this SHOULD have been what dawned on me first. I'm already using the `numpy` package to set the matrix. Why not use `numpy` tools to solve this?


I calculate the sum of the entire matrices terms and then subtract the sum of the diagonal terms from it. 

In [8]:
np.diag(A)  # this is more for arrays, but it still works.

array([1, 5, 9])

In [9]:
np.diagonal(A) # for matrices

array([1, 5, 9])

In [10]:
A.diagonal() # for matrices

matrix([[1, 5, 9]])

In [11]:
# SUM THE VALUES OF THE MATRIX AND THEN SUBTRACT OFF THE SUM OF THE DIAGONAL TERMS. 
np.sum(A) - np.sum(A.diagonal())

30

In [12]:
# SUM THE VALUES OF THE MATRIX AND THEN SUBTRACT OFF THE SUM OF THE DIAGONAL TERMS. 
np.sum(B) - np.sum(B.diagonal())

102

Oh yeah! This sparked a thought in my mind... the Trace of a Matrix is the sum of its diagonal terms!! I can replace `np.sum(A.diagonal())` with `np.trace(A)` or `A.trace()`.

In [13]:
# trace is the sum of the diag elements. 
np.sum(A) - np.trace(A)

30

In [14]:
A.sum() - A.trace()

matrix([[30]])

In [15]:
# trace is the sum of the diag elements. 
np.sum(B) - np.trace(B)

102

In [16]:
B.sum() - B.trace()

matrix([[102]])

In [17]:
# SET IT UP AS A FUNCTION.
sum_Off_Diags = lambda mat: np.sum(mat) - np.trace(mat)

sum_Off_Diags(A), sum_Off_Diags(B)

(30, 102)

```


```

__Other ways to solve this problem?__

1. Replace the diagonal terms with 0's and take the sum of the matrix.
2. Delete the diagonal terms and then take the sum of the matrix. 

<br>

__*!!! WARNING !!!:*__<br>
__*If you are going to alter a numpy matrix or array, such as replacing or deleting its values, make a copy of it first and then make those alterations on that copy. This is a good habit to get into, even when the function doesn't alter the original matrix or array, but creates a temporary one (i.e, is not kept by Python unless you instantiate it to a new variable).*__



In [18]:
# REPLACE DIAGONAL TERMS WITH 0'S AND TAKE THE SUM.
def sum_of_off_diags(mat):
    '''
    NOTE: must use np.sum(mat). 
    sum(mat) won't work.
    
    '''
    # MUST HAVE THIS OR YOU WILL ALTER THE ORIGINAL MATRIX PASSED TO THE FUNCTION!!
    mat = mat.copy() 
    
    n = len(mat)
    for i in range(0, n):
        mat[i,i] = 0
    return np.sum(mat)
    

In [19]:
sum_of_off_diags(A)

30

In [20]:
sum_of_off_diags(B)

102

In [21]:
# DELETE DIAGONAL TERMS AND TAKE THE SUM.
def sum_of_off_diags(mat):
    '''
    NOTE: must use np.sum(mat). 
    sum(mat) won't work.
    
    DELETE THE DIAGONAL TERMS FROM THE MATRIX.
    DOING IT THIS WAY RETURNS AN ARRAY OF THE OFF-DIAG TERMS.
    SIMPLY TAKE THE SUM THEN.
    
    '''
    # MUST HAVE THIS OR YOU WILL ALTER THE ORIGINAL MATRIX PASSED TO THE FUNCTION!!
    mat = mat.copy() 
    
    n = len(mat)
    to_delete = np.arange(0, (n**2)+1, n+1)
    # using np.delete doesn't change the matrix. Must instantiate to a new variable 
    # to save the array with deleted items. BUT, we don't need to, just need a sum.
    return np.sum(np.delete(mat, to_delete) )
    

In [22]:
sum_of_off_diags(A)

30

In [23]:
sum_of_off_diags(B)

102